<a href="https://colab.research.google.com/github/wldopie/ESAA/blob/main/HW_221007.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Chap 8. 텍스트 분석**
  - NLP와 텍스트 분석 비교
  1. `NLP`
    - 인간의 언어를 이해하고 해석하는 데 중점을 둠
    - 언어를 해석하기 위한 기계 번역, 자동으로 질문을 해석하고 답을 해주는 질의응답 시스템 등의 영역에서 텍스트 분석과 차별점
  2. `텍스트 분석`
    - 비정형 텍스트에서 의미 있는 정보를 추출하는 것에 중점을 둠
    - 머신러닝, 언어 이해, 통계 등을 활용해 모델을 수립하고 정보를 추출해 business intelligence나 예측 분석 등의 분석 작업을 주로 수행

  - 다양한 텍스트 분석 기법
    1. `텍스트 분류`(Text Classification)
      - 문서가 특정 분류 또는 카테고리에 속하는 것을 예측하는 기법
      - 특정 신문 기사 내용이 연애/정치/사회/문화 중 어떤 카테고리에 속하는지 자동으로 분류하거나, 스팸 메일 검출 같은 프로그램이 이에 속한다.
      - 지도학습을 적용한다.
    2. `감성 분석`(Sentiment Analysis)
      - 텍스트에서 나타나는 감정/판단/믿음/의견/기분 등의 주관적인 요소를 분석하는 기법
      - 소셜 미디어 감정 분석, 영화나 제품에 대한 긍정 또는 리뷰, 여론조사 의견 분석 등 다양한 분야에서 활용된다.
      - 지도학습 뿐만 아니라 비지도학습을 이용해 적용할 수 있다.
    3. `텍스트 요약`(Summarization)
      - 텍스트 내에서 중요한 주제나 중심 사상을 추출하는 기법
      - 대표적으로 토픽 모델링이 있다.
    4. `텍스트 군집화(Clustering)와 유사도 측정`
    - 비슷한 유형의 문서에 대해 군집화를 수행하는 기법
    - 비지도학습으로 수행하는 방법의 일환으로 사용된다.
    - 유사도 측정 역시 문서들간의 유사도를 측정해 비슷한 문서끼리 모을 수 있는 방법


---

## **8.1 텍스트 분석 이해**

### a. 텍스트 분석 수행 프로세스
  1. 텍스트 사전 준비 작업(텍스트 전처리)
    - 텍스트를 피처로 만들기 전에 미리 클렌징, 대/소문자 변경, 특수문자 삭제 등의 클렌징 작업, 단어 등의 토큰화 작업, 의미 없는 단어 제거 작업, 어근 추출 등의 텍스트 정규화 작업을 수행하는 것
  
  2. 피처 벡터화/추출
    - 사전 준비 작업으로 가공된 텍스트에서 피처를 추출하고 여기에 벡터 값을 할당한다. 
    - 대표적인 방법은 BOW와 Word2Vec이 있으며, BOW는 대표적으로 Count 기반과 TF-IDF 기반 벡터화가 있다.
  
  3. ML 모델 수립 및 학습/예측/평가
    - 피처 벡터화된 데이터 세트에 ML 모델을 적용해 학습/예측 및 평가를 수행

### b. 파이썬 기반의 NLP, 텍스트 분석 패키지
  1. NLTK(Natural Language Toolkit for Python)
    - 파이썬의 가장 대표적인 NLP 패키지
    - 방대한 데이터 세트와 서브 모듈을 가지고 있으며, NLP의 거의 모든 영역을 커버한다.
    - 수행 속도 측면에서는 아쉬운 부분이 있어 실제 대량의 데이터 기반에서는 활용이 어렵다.

  2. Gensim
    - 토픽 모델링 분야에서 가장 두각을 나타내는 패키지
    - 토픽 모델링을 쉽게 구현할 수 있는 기능을 제공해 왔으며, Word2Vec 구현 등의 다양한 신기능을 제공한다.

  3. SpaCy
    - 뛰어난 수행 성능으로 최근 가장 주목 받는 NLP 패키지
    - 많은 NLP 애플리케이션에서 SpaCy를 사용하는 사례가 늘고 있다.

---

## **8.2 텍스트 사전 준비 작업(텍스트 전처리) - 텍스트 정규화**
  - 텍스트 정규화 작업의 분류
    1. 클렌징(Cleansing)
    2. 토큰화(Tokenization)
    3. 필터링/스톱 워드 제거/철자 수정
    4. Stemming
    5. Lemmatization

  - 텍스트 정규화의 주요 작업을 NLTK 패키지를 이용해 실습해보자.

### a. 클렌징
  - 텍스트에서 분석에 오히려 방해가 되는 불필요한 문자, 기호 등을 사전에 제거하는 작업
  - HTML, XML 태그나 특정 기호 등을 사전에 제거

### b. 텍스트 토큰화
  - 토큰화의 유형
    1. 문서에서 문장을 분리하는 `문장 토큰화`
    2. 문서에서 단어를 토큰으로 분리하는 `단어 토큰화`

  - `문장 토큰화`
    - 문장의 마침표, 개행문자 등 문장의 마지막을 뜻하는 기호에 따라 분리하는 것이 일반적이다. 또한 정규 표현식에 따른 문장 토큰화도 가능하다.
    - NTLK에서 일반적으로 많이 쓰이는 `sent_tokenize`를 이용해 토큰화를 수행해보자.
    - 아래 코드에서 `nlt.download('punkt')`는 마침표, 개행 문자 등의 데이터 세트를 다운로드 받는다.

In [2]:
from nltk import sent_tokenize
import nltk
nltk.download('punkt')

text_sample = 'The Matrix is everywhere its all around us, here even in this room. \
You can see it out your window or on your television. \
You feel it when you go to work, or go to church or pay your taxes.'

sentences = sent_tokenize(text = text_sample)

print(type(sentences), len(sentences))
print(sentences)

<class 'list'> 3
['The Matrix is everywhere its all around us, here even in this room.', 'You can see it out your window or on your television.', 'You feel it when you go to work, or go to church or pay your taxes.']


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


  - sent_tokenize()가 반환하는 것은 각각의 문장으로 구성된 list 객체이다.
  - 반환된 list 객체가 3개의 문장으로 된 문자열을 가지고 있는 것을 알 수 있다.

  - `단어 토큰화`
    - 문장을 단어로 토큰화하는 것
    - 기본적으로 공백, 콤마, 마침표, 개행문자 등으로 단어를 분리하지만, 정규 표현식을 이용해 다양한 유형으로 토큰화를 수행
    - NLTK에서 기본으로 제공하는 `word_tokenize()`를 이용해 단어로 토큰화해보자.

In [3]:
from nltk import word_tokenize

sentence = "The Matrix is everywhere its all around us, here even in this room."
words = word_tokenize(sentence)
print(type(words), len(words))
print(words)

<class 'list'> 15
['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.']


  - 이번에는 `sent_tokenize` `word_tokenize` 조합해 문서에 대해 모든 단어를 토큰화해보자.
  - 이전 예제에서 선언된 3개의 문장으로 된 text_sample을 문장별로 단어 토큰화를 적용한다.
  - 이를 위해 문서를 먼저 문장으로 나누고, 개별 문장을 다시 단어로 토큰화하는 `tokenize_text()`함수를 생성하자.

In [5]:
from nltk import word_tokenize, sent_tokenize

# 여러 개의 문장으로 된 입력 데이터를 문장별로 단어 토큰화하게 만드는 함수 생성
def tokenize_text(text):

  # 문장별로 분리 토큰
  sentences = sent_tokenize(text)
  # 분리된 문장별 단어 토큰화
  word_tokens = [word_tokenize(sentence) for sentence in sentences]
  return word_tokens

# 여러 문장에 대해 문장별 단어 토큰화 수행
word_tokens = tokenize_text(text_sample)
print(type(word_tokens), len(word_tokens))
print(word_tokens)

<class 'list'> 3
[['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.'], ['You', 'can', 'see', 'it', 'out', 'your', 'window', 'or', 'on', 'your', 'television', '.'], ['You', 'feel', 'it', 'when', 'you', 'go', 'to', 'work', ',', 'or', 'go', 'to', 'church', 'or', 'pay', 'your', 'taxes', '.']]


  - 위와 같이 문장을 단어별로 하나씩 토큰화 할 경우, 문맥적인 의미는 무시된다.
  - 이를 해결해 보고자 도입된 것이 n-gram이다.
  - `n-gram`
    - 연속된 n개의 단어를 하나의 토큰화 단위로 분리해 내는 것
    - n개 단어 크기 윈도우를 만들어 문장의 처음부터 오른쪽으로 움직이면서 토큰화를 수행


### c. 스톱 워드 제거
  - 스톱 워드(Stop word)
    - 분석에 큰 의미가 없는 단어
  - 문법적으로는 중요하지만 문맥적으로는 큰 의미가 없는 단어들의 경우, 빈번하게 텍스트에 나타나므로 이들을 사전에 제거하지 않으면 그 빈번함으로 인해 중요한 단어로 인지될 수 있다. 따라서 이 의미 없는 단어를 제거하는 것이 중요한 전처리 작업이다.
  - 언어별로 이러한 스톱 워드가 목록화돼 있다. 
  - NLTK의 경우 가장 다양한 언어의 스톱 워드를 제공한다.

In [6]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [7]:
print('영어 stop words 개수:', len(nltk.corpus.stopwords.words('english')))
print(nltk.corpus.stopwords.words('english')[:20])

영어 stop words 개수: 179
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his']


  - 영어의 경우 스톱 워드의 개수가 179개이며, 그 중 20개만 살펴보면 위의 결과와 같다.
  - 바로 위 예제에서 3개의 문장별로 단어를 토큰화해 생성된 word_tokens 리스트에 대해서 stopwords를 필터링으로 제거해 분석을 위한 의미 있는 단어만 추출해보자.

In [8]:
import nltk

stopwords = nltk.corpus.stopwords.words('english')
all_tokens = []

# 위 예제에서 3개의 문장별로 얻은 word_tokens list에 대해 스톱 워드를 제거하는 반복문
for sentence in word_tokens:
  filtered_words = []
  # 개별 문장별로 토큰화된 문장 list에 대해 스톱 워드를 제거하는 반복문
  for word in sentence:
    # 소문자로 모두 변환
    word = word.lower()
    # 토큰화된 개별 단어가 스톱 워드의 단어에 포함되지 않으면 word_tokens에 추가
    if word not in stopwords:
      filtered_words.append(word)
  all_tokens.append(filtered_words)

print(all_tokens)

[['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['see', 'window', 'television', '.'], ['feel', 'go', 'work', ',', 'go', 'church', 'pay', 'taxes', '.']]


  - is, this와 같은 스톱 워드가 필터링을 통해 제거됐음을 알 수 있다.

### d. Stemming과 Lemmatization
  - 문법적 또는 의미적으로 변화하는 단어의 원형을 찾는 것
  - Lemmatization이 Stemming보다 정교하며, 의미론적인 기반에서 단어의 원형을 찾는다.
  - Stemming
    - 원형 단어로 변환 시 일반적인 방법을 적용하거나 더 단순화된 방법을 적용해 원래 단어에서 일부 철자가 훼손된 어근 단어를 추출
  - Lemmatization
    - 품사와 같은 문법적인 요소와 더 의미적인 부분을 감안해 정확한 철자로 된 어근 단어를 찾아준다.
  - 따라서 Lemmatization이 Stemming보다 변환에 더 오랜 시간을 필요로 한다.
  - NLTK에서 다양한 클래스를 이용해 Stemming과 Lemmatization을 비교해보자.

In [9]:
from nltk.stem import LancasterStemmer
stemmer = LancasterStemmer()

print(stemmer.stem('working'), stemmer.stem('works'), stemmer.stem('worked'))
print(stemmer.stem('amusing'), stemmer.stem('amuses'), stemmer.stem('amused'))
print(stemmer.stem('happier'), stemmer.stem('happiest'))
print(stemmer.stem('fancier'), stemmer.stem('fanciest'))

work work work
amus amus amus
happy happiest
fant fanciest


  - work의 경우 진행형, 3인칭 단수, 과거형 모두 기본 단어인 work에 ing, s, ed가 붙는 단순한 변화이므로, 원형 단어로 work를 제대로 인식한다.
  - 그러나 amuse의 경우, 각 변화가 amuse가 아닌 amus에 ing, s, ed가 붙으므로 정확한 단어인 amuse가 아닌 amus를 원형으로 인식한다. 
  - 형용사인 happy, fancy의 경우도 비교형, 최상급형으로 변형된 단어의 정확한 원형을 찾지 못하고 원형 단어에서 철자가 다른 어근 단어로 인식하는 경우가 발생한다.

  - 이번에는 `WordNetLemmatizer`를 이용해 Lemmatization을 수행해보자.
  - 일반적으로 Lemmatization은 보다 정확한 원형 단어 추출을 위해 단어의 '품사'를 입력해줘야 한다.
  - 동사의 경우 v, 형용사의 경우 a를 입력한다.

In [13]:
from nltk.stem import WordNetLemmatizer
nltk.download('all')

lemma = WordNetLemmatizer()
print(lemma.lemmatize('amusing', 'v'), lemma.lemmatize('amuses', 'v'), lemma.lemmatize('amused', 'v'))
print(lemma.lemmatize('happier', 'a'), lemma.lemmatize('happiest', 'a'))
print(lemma.lemmatize('fancier', 'a'), lemma.lemmatize('fanciest', 'a'))

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping grammars/basque_grammars.zip.
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/biocreative_ppi.zip.
[nltk_data]    | Downloadin

amuse amuse amuse
happy happy
fancy fancy


  - 앞의 Stemmer보다 정확하게 원형 단어를 추출해줌을 알 수 있다.

---

## **8.3 Bag of Words - BOW**
  - Bag of Words(BOW)
    - 문서가 가지는 모든 단어를 문맥이나 순서를 무시하고 일괄적으로 단어에 대해 빈도 값을 부여해 피처 값을 추출하는 모델
    - 문서 내 모든 단어를 한꺼번에 봉투 안에 넣은 뒤에 흔들어서 섞는다는 의미를 갖는다.

  - BOW 모델의 `장점`
    1. 쉽고 빠른 구축이 가능하다.
    2. 예상보다 문서의 특징을 잘 나타낼 수 있는 모델이어서 전통적으로 여러 분야에서 활용도가 높다.

  - BOW 모델의 `단점`
    1. 문맥 의미 반영 부족
      - BOW는 단어의 순서를 고려하지 않기 때문에 문장 내에서 단어의 문맥적인 의미가 무시된다. 이를 보완하기 위해 n_gram 기법을 사용할 수 있지만, 제한적인 부분에 그치므로 언어의 많은 부분을 차지하는 문맥적인 해석을 처리하지 못한다.
    2. 희소 행렬 문제
      - BOW로 피처 벡터화를 수행하면 희소 행렬 형태의 데이터 세트가 만들어지기 쉽다.
      - 많은 문서에서 단어를 추출하면 매우 많은 단어가 칼럼으로 만들어진다. 문서마다 서로 다른 단어로 구성되므로, 단어가 문서마다 나타나지 않는 경우가 훨씬 많다.
      - 즉, 매우 많은 문서에서 수십만 개의 단어가 사용되는데, 하나의 문서에 있는 단어는 극히 일부이므로 대부분의 데이터가 0으로 채워지게 된다.
      - 이처럼 대규모의 칼럼으로 구성된 행렬에서 대부분의 값이 0으로 채워지는 행렬을 희소 행렬(Sparse Matrix)이라고 한다. 
      - 이와는 반대로 대부분의 값이 0이 아닌 의미 있는 값으로 채워져 있는 행렬을 밀집 행렬(Dense Matrix)이라고 한다.
      - 희소 행렬은 일반적으로 알고리즘의 수행 시간과 예측 성능을 떨어뜨리기 때문에, 희소 행렬을 위한 특별한 기법이 마련돼 있다.

### a. BOW 피처 벡터화
  - 피처 벡터화
    - 각 문서의 텍스트를 단어로 추출해 피처로 할당하고, 각 단어의 발생 빈도와 같은 값을 이 피처에 값으로 부여해 각 문서를 이 단어 피처의 발생 빈도 값으로 구성된 벡터로 만드는 기법

  - BOW 모델에서의 피처 벡터화
    - 모든 문서에서 모든 단어를 칼럼 형태로 나열하고 각 문서에서 해당 단어의 횟수나 정규화된 빈도를 값으로 부여하는 데이터 세트 모델로 변경하는 것

  - BOW 피처 벡터화의 2가지 종류
    1. 카운트 기반의 벡터화
      - 단어 피처에 값을 부여할 때 각 문서에서 해당 단어가 나타나는 횟수, 즉 count를 부여하는 경우
      - 카운트 값이 높을수록 중요한 단어로 인식
      - 그러나 카운트만 부여할 경우, 그 문서의 특징을 나타내기보다는 언어의 특성상 문장에서 자주 사용될 수 밖에 없는 단어까지 높은 값을 부여 
    2. TF-IDF(Term Frequency - Inverse Document Frequency) 기반의 벡터화
      - 개별 문서에서 자주 나타나는 단어에 높은 가중치를 두되, 모든 문서에서 전반적으로 자주 나타나는 단어에 대해서는 페널티를 주는 방식으로 값을 부여
      - 모든 문서에서 반복적으로 자주 발생하는 단어에 대해서는 페널티를 부여하는 방식으로 단어에 대한 가중치를 균형을 맞추는 것
      - 문서마다 텍스트가 길고 문서의 개수가 많은 경우 카운트 방식보다는 TF-IDF 방식을 사용하는 것이 더 좋은 예측 성능을 보장할 수 있다.

### b. 사이킷런의 Count 및 TF-IDF 벡터화 구현: `CountVectorizer`, `TfidfVectorizer`
  - `CountVectorizer` 클래스
    - 카운트 기반의 벡터화를 구현한 클래스
    - 피처 벡터화, 소문자 일괄 변환, 토큰화, 스톱 워드 필터링 등의 텍스트 전처리도 함께 수행
    - 파라미터
      - `max_df` : 너무 높은 빈도수를 가지는 단어 피처를 제외하기 위한 파라미터
      - `min_df` : 너무 낮은 빈도수를 가지는 단어 피처를 제외하기 위한 파라미터
      - `max_features` : 추출하는 피처의 개수를 제한
      - `stop_words` : 스톱 워드로 지정된 단어는 추출에서 제외
      - `n_gram_range_` : BOW 모델의 단어 순서를 보강
      - `analyzer` : 피처 추출을 수행할 단위를 지정
      - `token_pattern` : 토큰화를 수행하는 정규 표현식 패턴을 지정
      - `tokenizer` : 토큰화를 별도의 커스텀 함수로 이용시 적용

  - CountVectorizer 클래스를 이용한 텍스트의 피처 벡터화 방법
    1. 모든 문자를 소문자로 변경하는 등의 전처리 작업 수행
    2. 디폴트로 단어 기준으로 n_gram_range를 반영해 각 단어를 토큰화
    3. 텍스트 정규화를 수행
    4. 마지막으로 max_df, min_df, max_feature 등의 파라미터를 이용해 토큰화된 단어를 피처로 추출하고 단어 빈도수 벡터 값을 적용

  - `TfidfVectorizer` 클래스를 이용해 TF-IDF 벡터화 진행

### c. BOW 벡터화를 위한 희소 행렬
  - 희소 행렬은 너무 많은 불필요한 0 값이 메모리 공간에 할당되어 메모리 공간이 많이 필요하며, 행렬의 크기가 커서 연산 시에도 데이터 엑세스를 위한 시간이 많이 소모된다.
  - 따라서 이러한 `희소 행렬을 물리적으로 적은 메모리 공간을 차지할 수 있도록 변환`해야 하는데, 대표적인 방법으로 `COO 형식`과 `CSR 형식`이 있다.

### d. 희소 행렬 - COO 형식
  - 0이 아닌 데이터만 별도의 데이터 배열에 저장하고, 그 데이터가 가리키는 행과 열의 위치를 별도의 배열로 저장하는 방식
  - 희소 행렬 변환을 위해 주로 Scipy를 사용한다.
    - Scipy의 sparse를 이용해 희소 행렬 변환을 COO 형식으로 수행해보자.

In [14]:
import numpy as np

dense = np.array([ [3, 0, 1], [0, 2, 0] ])

  - 위의 밀집 행렬을 Scipy의 `coo_matrix` 클래스를 이용해 COO 형식의 희소 행렬로 변환해보자.

In [15]:
from scipy import sparse

# 0이 아닌 데이터 추출
data = np.array([3, 1, 2])

# 행 위치와 열 위치를 각각 배열로 생성
row_pos = np.array([0, 0, 1])
col_pos = np.array([0, 2, 1])

# sparse 패키지의 coo_matrix를 이용해 COO 형식으로 희소 행렬 생성
sparse_coo = sparse.coo_matrix((data, (row_pos, col_pos)))

  - 이를 `toarray()` 메서드를 이용해 다시 밀집 형태의 행렬로 출력해보자.

In [16]:
sparse_coo.toarray()

array([[3, 0, 1],
       [0, 2, 0]])

### e. 희소 행렬 - CSR 형식
  - COO 형식이 행과 열의 위치를 나타내기 위해서 반복적인 위치 데이터를 사용해야 하는 문제점을 해결한 방식
  - 행 위치 배열 내에 있는 고유한 값의 시작 위치만 다시 별도의 위치 배열로 가지는 변환 방식
  - 고유 값의 시작 위치만 알고 있으면 얼마든지 행 위치 배열을 다시 만들 수 있어 COO 방식보다 메모리가 적게 들고 빠른 연산이 가능하다.  

In [17]:
from scipy import sparse

dense2 = np.array([[0,0,1,0,0,5],
                   [1,4,0,3,2,5],
                   [0,6,0,3,0,0],
                   [2,0,0,0,0,0],
                   [0,0,0,7,0,8],
                   [1,0,0,0,0,0]])

# 0이 아닌 데이터 추출
data2 = np.array([1,5,1,4,3,2,5,6,3,2,7,8,1])

# 행 위치와 열 위치를 각각 array로 생성
row_pos = np.array([0,0,1,1,1,1,1,2,2,3,4,4,5])
col_pos = np.array([2,5,0,1,3,4,5,1,3,0,3,5,0])

# COO 형식으로 변환
sparse_coo = sparse.coo_matrix((data2, (row_pos, col_pos)))

# 행 위치 배열의 고유한 값의 시작 위치 인덱스를 배열로 생성
row_pos_ind = np.array([0,2,7,9,10,12,13])

# CSR 형식으로 변환
sparse_csr = sparse.csr_matrix((data2, col_pos, row_pos_ind))

print('COO 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인')
print(sparse_coo.toarray())
print('CSR 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인')
print(sparse_csr.toarray())

COO 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]
CSR 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]


  - 실제 사용 시에는 다음과 같이 밀집 행렬을 생성 파라미터로 입력하면 COO나 CSR 희소 행렬로 생성한다.

In [18]:
dense3 = np.array([[0,0,1,0,0,5],
                   [1,4,0,3,2,5],
                   [0,6,0,3,0,0],
                   [2,0,0,0,0,0],
                   [0,0,0,7,0,8],
                   [1,0,0,0,0,0]])

coo = sparse.coo_matrix(dense3)
csr = sparse.csr_matrix(dense3)